File: extra-code/stage2_code/cpp_srv/src/add_two_ints_client.cpp
// ============================================================
// Service 通信示例:客户端(Client,面向对象 + 异步回调写法)
// 演示:连续发出 3 个 AddTwoInts 请求,用回调异步接收结果,全部完成后自动退出
// 运行:先启动 add_two_ints_server,再 ros2 run cpp_srv add_two_ints_client
// 对比:add_two_ints_client_no_oop.cpp 是更简单的"发一次等一次"写法
// ============================================================

#include <chrono>
#include <exception>
#include <memory>

#include "example_interfaces/srv/add_two_ints.hpp"
#include "rclcpp/rclcpp.hpp"

// 引入时间字面量,使 5s 表示 5 秒
using namespace std::chrono_literals;

class AddTwoIntsClient : public rclcpp::Node
{
public:
  AddTwoIntsClient() : Node("add_two_ints_client")
  {
    // create_client 只是创建客户端句柄,并不保证服务端已经在线
    client_ = create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
    pending_requests_ = 0;
  }

  // 发请求前先确认服务端在线:wait_for_service 最多阻塞 5 秒,超时返回 false
  bool waitForService()
  {
    if (client_->wait_for_service(5s)) {
      return true;
    }
    RCLCPP_ERROR(get_logger(), "/add_two_ints service is not available");
    return false;
  }

  // 异步发送一个请求:async_send_request 立即返回不等结果,
  // 结果到达时由 spin 的事件循环调用我们注册的 lambda 回调
  void callAddTwoInts(int64_t a, int64_t b)
  {
    auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
    request->a = a;
    request->b = b;
    pending_requests_++;
    // 第二个参数是 lambda 回调,[this, request] 是捕获列表:
    // 捕获 this 才能在 lambda 里用成员变量,捕获 request 是为了打印时还能拿到原始参数
    // 回调参数 SharedFuture 代表"将来才有的结果",future.get() 取出 Response
    client_->async_send_request(
      request,
      [this, request](rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedFuture future) {
        try {
          RCLCPP_INFO(get_logger(), "%ld + %ld = %ld", request->a, request->b, future.get()->sum);
        } catch (const std::exception & error) {
          RCLCPP_ERROR(get_logger(), "Service call failed: %s", error.what());
        }
        // 用计数器跟踪未完成的请求,全部收到回复后主动 shutdown 让 main 里的 spin 退出
        pending_requests_--;
        if (pending_requests_ == 0) {
          RCLCPP_INFO(get_logger(), "All requests completed.");
          rclcpp::shutdown();
        }
      });
  }

private:
  rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedPtr client_;
  int pending_requests_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<AddTwoIntsClient>();
  if (!node->waitForService()) {
    rclcpp::shutdown();
    return 1;
  }
  // 三个请求几乎同时发出(异步的优势:不用发一个等一个)
  node->callAddTwoInts(2, 7);
  node->callAddTwoInts(1, 4);
  node->callAddTwoInts(10, 20);
  // 必须 spin 才能收到回复:回调都在事件循环中执行
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: extra-code/stage2_code/py_srv/py_srv/add_two_ints_client.py
# ============================================================
# Service 通信示例：异步客户端（OOP 写法）
# 演示概念：call_async 异步调用、Future 回调、多请求并发
# 连续发 3 个请求而不阻塞等待，收齐全部响应后自动退出
# 先启动服务端，再运行：ros2 run py_srv add_two_ints_client
# ============================================================
# partial 用于给回调函数预先绑定额外参数（这里是 request）
from functools import partial

import rclpy
from example_interfaces.srv import AddTwoInts
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node


class AddTwoIntsClient(Node):
    """Asynchronous OOP AddTwoInts client that exits after all replies arrive."""

    def __init__(self):
        super().__init__("add_two_ints_client")
        # 创建 Client：接口类型和服务名必须与服务端一致
        self.client_ = self.create_client(AddTwoInts, "add_two_ints")
        # 记录未完成的请求数，用于判断何时全部完成
        self.pending_requests_ = 0

    def wait_for_service(self):
        # 先确认服务端在线（最多等 5 秒），
        # 否则请求会一直挂着得不到响应
        if self.client_.wait_for_service(timeout_sec=5.0):
            return True
        self.get_logger().error("/add_two_ints service is not available")
        return False

    def call_add_two_ints(self, a, b):
        request = AddTwoInts.Request()
        request.a = a
        request.b = b
        # call_async 不阻塞：立即返回一个 Future（未来结果的
        # 占位符），响应到达后由 spin 触发 done 回调
        future = self.client_.call_async(request)
        self.pending_requests_ += 1
        # 用 partial 把 request 一起传进回调，
        # 这样打印结果时能知道对应的是哪个请求
        future.add_done_callback(
            partial(self.callback_call_add_two_ints, request=request))

    def callback_call_add_two_ints(self, future, request):
        # Future 完成时被调用：result() 取出响应，
        # 若服务端出错则会抛异常
        try:
            response = future.result()
            self.get_logger().info(f"{request.a} + {request.b} = {response.sum}")
        except Exception as exc:
            self.get_logger().error(
                f"Service call {request.a} + {request.b} failed: {exc}")
        finally:
            # 所有请求都完成后主动 shutdown，让 spin 退出
            self.pending_requests_ -= 1
            if self.pending_requests_ == 0:
                self.get_logger().info("All requests completed.")
                rclpy.shutdown()


def main(args=None):
    rclpy.init(args=args)
    node = AddTwoIntsClient()
    try:
        if node.wait_for_service():
            # 连续发出 3 个异步请求（互不等待），
            # 然后 spin 等待响应回调逐个触发
            node.call_add_two_ints(2, 7)
            node.call_add_two_ints(1, 4)
            node.call_add_two_ints(10, 20)
            rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()
